File: file_server.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');
const os = require('os');
const { spawn } = require('child_process');

const PORT = 3456;

const server = http.createServer((req, res) => {
  const parsedUrl = url.parse(req.url, true);
  const pathname = parsedUrl.pathname;

  console.log(`${req.method} ${req.url}`);

  if (req.method === 'POST') {
    // Dynamic POST /<folder>/<requestName> — validate XML against XSD
    const segments = pathname.replace(/^\/+/, '').split('/').filter(Boolean);
    if (segments.length < 2) {
      res.writeHead(404, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'Not found' }));
      return;
    }

    const folder = segments[0];
    const requestName = segments.slice(1).join('/');
    const xsdPath = path.join(__dirname, 'schemas', folder, `${requestName}.xsd`);

    // Check if the XSD file exists
    if (!fs.existsSync(xsdPath)) {
      res.writeHead(404, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'XSD not found', xsdPath }));
      return;
    }

    let body = '';
    req.on('data', (chunk) => {
      body += chunk.toString();
    });
    req.on('end', () => {
      // Write XML body to a temp file
      const tempXmlPath = path.join(os.tmpdir(), `validate_${Date.now()}.xml`);
      fs.writeFileSync(tempXmlPath, body);

      const validateScript = path.join(__dirname, 'validate.py');
      const child = spawn('python3', [validateScript, xsdPath, tempXmlPath], {
        cwd: __dirname
      });

      let combinedOutput = '';

      child.stdout.on('data', (data) => {
        combinedOutput += data.toString();
      });

      child.stderr.on('data', (data) => {
        combinedOutput += data.toString();
      });

      child.on('close', (exitCode) => {
        // Clean up temp file
        try { fs.unlinkSync(tempXmlPath); } catch (_) {}

        if (exitCode === 0 && combinedOutput.includes('✅')) {
          res.writeHead(200, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ valid: true, message: combinedOutput.trim() }));
        } else {
          res.writeHead(400, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ valid: false, message: combinedOutput.trim() }));
        }
      });

      child.on('error', (err) => {
        // Clean up temp file on error too
        try { fs.unlinkSync(tempXmlPath); } catch (_) {}
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: err.message }));
      });
    });
  } else {
    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Not found' }));
  }
});

server.listen(PORT, () => {
  console.log(`File server listening on http://localhost:${PORT}`);
});

File: validation_server/file_server.py
"""
file_server.py — HTTP validation server for XSD/XML validation.

Exposes a single endpoint that Postman pre-request scripts call to validate
the XML body of a request before it is sent to the real API.

Endpoint:
    POST /<folder>/<RequestName>

    Resolves the XSD at: schemas/<folder>/<RequestName>.xsd (relative to repo root)
    Reads the raw XML from the request body and validates it against that schema.

Responses:
    200  {"valid": true}
    400  {"valid": false, "message": "<validation error>"}
    404  {"error": "XSD not found", "xsdPath": "<path>"}
    500  {"error": "<message>"}

Usage:
    python validation_server/file_server.py
    # Listening on http://localhost:3456

Install dependencies first: pip install -r validation_server/requirements.txt
"""

import json
import os
import xmlschema
from http.server import BaseHTTPRequestHandler, HTTPServer

PORT = 3456
ROOT = os.path.join(os.path.dirname(__file__), '..')


class Handler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        """Override default logging to print method + path only."""
        print(f"{self.command} {self.path}")

    def send_json(self, status, data):
        """Write a JSON response with the given HTTP status code."""
        body = json.dumps(data).encode()
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', len(body))
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self):
        """Handle POST /<folder>/<RequestName> — validate XML body against XSD.

        Path segments map directly to the schema path:
            POST /sample/BasicDocument
            → schemas/sample/BasicDocument.xsd
        """
        segments = [s for s in self.path.split('/') if s]
        if len(segments) < 2:
            self.send_json(404, {'error': 'Not found'})
            return

        xsd_path = os.path.join(ROOT, 'schemas', segments[0], f"{'/'.join(segments[1:])}.xsd")
        if not os.path.exists(xsd_path):
            self.send_json(404, {'error': 'XSD not found', 'xsdPath': xsd_path})
            return

        length = int(self.headers.get('Content-Length', 0))
        xml_body = self.rfile.read(length).decode()

        try:
            xmlschema.XMLSchema(xsd_path).validate(xml_body)
            self.send_json(200, {'valid': True})
        except xmlschema.XMLSchemaValidationError as e:
            self.send_json(400, {'valid': False, 'message': str(e)})
        except Exception as e:
            self.send_json(500, {'error': str(e)})


if __name__ == '__main__':
    httpd = HTTPServer(('', PORT), Handler)
    print(f'File server listening on http://localhost:{PORT}')
    httpd.serve_forever()
